### PR TITLE
Extension cases sync

### PR DESCRIPTION
**Pull request by proteusvacuum**
Opening for tests / review. This is still in the preliminary stages...
@czue would be great if you could give this a glance (should be easiest commit by commit)

Todo:
- [ ] Fix skipped tests in `.json` shared test file
- [ ] Test subsequent syncs still do the right thing (e.g. delete, delegate, close, etc. still sends the correct cases)
- [ ] Make sure ownership cleanliness gets set properly with extensions

### CODE REVIEW

File: corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py [hunk 1]
```diff
@@ -20,6 +20,7 @@
 from casexml.apps.phone.restore import CachedResponse, RestoreConfig, RestoreParams, RestoreCacheSettings
 from casexml.apps.case.xml import V2, V1
 from casexml.apps.case.sharedmodels import CommCareCaseIndex
+from casexml.apps.case.const import CASE_INDEX_EXTENSION
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py]: F401 'CASE_INDEX_EXTENSION' imported but unused

File: corehq/ex-submodules/casexml/apps/phone/models.py [hunk 2]
```diff
@@ -472,6 +473,41 @@ class IndexTree(DocumentSchema):
     def __repr__(self):
         return json.dumps(self.indices, indent=2)
 
+    @staticmethod
+    def get_all_dependencies(case_id, child_index_tree, extension_index_tree,
```
Comment (proteusvacuum) [inline corehq/ex-submodules/casexml/apps/phone/models.py]: This is the crux of it. Still not convinced that this completely covers the spec (cf. skipped tests).

**Comment by czue** [comment]
wow, this is a beast! looks great overall and love all the tests. I didn't review those too closely - let me know if you think I should.

**Comment by dimagimon** [comment]
There are 9 errors:

* corehq/ex-submodules/casexml/apps/case/util.py, line 131 - E302 expected 2 blank lines, found 1
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 4 - F401 'CASE_INDEX_EXTENSION' imported but unused
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 5 - F401 'get_reverse_indices_for_case_id' imported but unused
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 9 - F401 'get_extension_cases' imported but unused
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 9 - F401 'get_indexed_cases' imported but unused
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 134 - W291 trailing whitespace
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 146 - E501 line too long (131 > 115 characters)
* corehq/ex-submodules/casexml/apps/phone/tests/test_cleanliness.py, line 642 - E501 line too long (118 > 115 characters)
* corehq/ex-submodules/casexml/apps/phone/tests/test_extension_indexes.py, line 105 - E501 line too long (117 > 115 characters)

**Comment by proteusvacuum** [comment]
(just wanted to run tests again, will clean up history and ping when ready for review again)

**Comment by dimagimon** [comment]
There are 9 errors:

* corehq/ex-submodules/casexml/apps/case/util.py, line 131 - E302 expected 2 blank lines, found 1
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 4 - F401 'CASE_INDEX_EXTENSION' imported but unused
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 5 - F401 'get_reverse_indices_for_case_id' imported but unused
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 9 - F401 'get_extension_cases' imported but unused
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 9 - F401 'get_indexed_cases' imported but unused
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 134 - W291 trailing whitespace
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 146 - E501 line too long (131 > 115 characters)
* corehq/ex-submodules/casexml/apps/phone/tests/test_cleanliness.py, line 642 - E501 line too long (118 > 115 characters)
* corehq/ex-submodules/casexml/apps/phone/tests/test_extension_indexes.py, line 105 - E501 line too long (117 > 115 characters)

**Pull request closed without merging**